### ISSUE TITLE
Blazor should implement ways to manipulate dom in another path rather than component rendering.

### ISSUE THREAD

**Issue by Xyncgas** [dotnet/aspnetcore#33598]
Webassembly can't directly manipulate dom, and in the future while webassembly is a medium for languages to compile to the web it can be running in many environments too that might be without implementing a dom environment.
Blazor is a product microsoft makes that builds on top of webassembly along with other technologies of the .net stacks, we have server and webassembly blazor now, I see the possibility that each of them can implement the ability to manipulate dom in another way rather than component rendering :eg. Javascript api wrapped inside blazor. Javascript is a moving target but as it is a main environment that blazor runs it might as well chase after it.
The need for this is that, through current methods of manipulating dom through delta rendering, I can't trigger animations by element's css attributes like 'animation' and 'transition', because the browsers seems to think it becomes another different element that you are replacing instead of actually **'manipulating**.
1) Either implement the javascript api wrapper since it's generally helpful to be able to use javascript through intellisense yet completely in c# context.
2) Signal browsers to treat changes in dom by blazor's component rendering as actually changing the element instead of replacing it with document context.
3) Or since u can just do things on you own, instead of replacing the dom with calculated delta, how about actually take the delta to the dom and compare changes and change the element ITSELF. For example instead of replacing
```
<div>
*Context_to_be_re-rendered
</div>
```
with :
```
<div>
*rendered_new_context
</div>
```
You expand the `*Context_to_be_re-rendered` and either remove the element since delta shows it doesn't exist while having no child nodes, or change the attribute and tag name and stuff to the delta.

This the problem I've been having with blazor's rendering mechanisms too, it might be easier to just replace the element document with the delta, it makes my life more difficult because the browsers now consider them as something else instead of changes made on the original context, it is also not difficult to do this, I am also not asking a breaking change instead it's another feature because currently I have no ways to trick browsers thinking I am just changing some inline styles of the html element/node without using javascript I hand written myself. I could do this all day it's just that I feel like this should be part of blazor's features like I would even do it for free if I am good enough just to put it in blazor

**Comment by javiercn** [comment]
@Xyncgas thanks for contacting us.

1) We don't plan to change the way Blazor renders on the page. If we change that, it stops being Blazor and becomes a new framework.
2) We already calculate the minimal diff to the DOM and update existing nodes.
3) Our rendering process is similar to that of other SPA frameworks and they have found solutions for doing animations, so what works for them should work for Blazor too.

Either we are not understanding correctly what you are asking of us (and if that's the case please provide a set of concrete examples where the current behavior is problematic) or you are asking for fundamental changes to the framework that affect the core programming model and that would require a huge investment for unclear payoffs, which is not something we are willing to tackle.

**Comment by Xyncgas** [comment]
Another blocking reason, I as many people are trying to make good looking, responsive websites, sometimes I detect width and height of viewport.
Now as parameter of the components, it's pretty convenient that I can manipulate dom to adjust content. At least it should be that way.
Because the way blazor refreshes components when parameter change by re-rendering the dom in a way that would make browser think it is a different element in the dom now, omg !!! Now my javascript variables that are holding a reference to the element(document.getElementById()) is going to be null and make my functions fail and I just wanted to change its width, while my component's css animation is going to start playing again which is annoying when user simply rotated the screen.
Please offer alternatives to re-writing deltas directly to dom and change it through javascript actually 'maniulating' the elements please.

**Comment by Xyncgas** [comment]
ok let me get you something to try for yourself.

**Comment by Xyncgas** [comment]
I am so sleepy it's night time and I've got work let me come back to this another day thanks

**Comment by Xyncgas** [comment]
I am going to investigate further as the problem seemingly fixed itself after I did something last night n compiled again in the morning

**Comment by Xyncgas** [comment]
anyways here's the code

**Comment by Xyncgas** [comment]
```
@inject IJSRuntime JS
@code{
    [Parameter]
    public float Observable_screenSize_reference { get; set; }
    [Parameter]
    public float window_innerHeight { get; set; }
    [Parameter]
    public float window_innerWidth { get; set; }
    bool Showing = true;
}
@code
{
    System.Timers.Timer Timer = null;
    static readonly int Duration = 2000;//in milliseconds
    static readonly int stall = 500;//in milliseconds
    void Destruction_after_duration()
    {
        if (Timer == null)
        {
            Timer = new System.Timers.Timer(Duration + stall + 100);//after stall (giving it a breathing room for dom to construct), animation happen, transition lasts $duration
            Timer.Elapsed += ((Object source, System.Timers.ElapsedEventArgs e) =>
            {
                Showing = false;
                StateHasChanged();
            });
            Timer.AutoReset = false;
            Timer.Enabled = true;
        }
    }
}
@code{
    Layout_algorithm.Element_id element_Id = new Layout_algorithm.Element_id(9);
    string Initial_height;
    string Initial_width;
    string margin_ = string.Empty; bool enabled_ = true;
    void Initialize_dimension_after_first_render()
    {
        var a_portion_of_the_ground = (Observable_screenSize_reference * 0.2F).ToString() + "px";
        Initial_height = a_portion_of_the_ground;
        Initial_width = a_portion_of_the_ground;
    }
    System.Timers.Timer Animation_timer_stalling_for_latency=null;//ultimately I should be using css @keyframe
    void Animation_expansion()
    {
        if (display_conditions_met())
        {
            if(Animation_timer_stalling_for_latency==null)
            {
                Animation_timer_stalling_for_latency = new System.Timers.Timer(stall);
#if true
                Animation_timer_stalling_for_latency.Elapsed += (async (Object source, System.Timers.ElapsedEventArgs e) =>
                {
                    Initial_height = "100vh";
                    Initial_width = "100vw";
                    margin_ = "0"; enabled_ = false;
                    StateHasChanged();
                });
#else
                Animation_timer_stalling_for_latency.Elapsed += (async (Object source, System.Timers.ElapsedEventArgs e) =>
                {
                    var ta = Javascript_api_wrapper.css_styles.widthById(JS, element_Id.get(0), "100vw");
                    var tb = Javascript_api_wrapper.css_styles.heightById(JS, element_Id.get(0), "100vh");
                    var tc = Javascript_api_wrapper.css_styles.marginById(JS, element_Id.get(0), "0");
                    await ta; await tb; await tc;
                });
#endif//one uses javascript api one without
                Animation_timer_stalling_for_latency.Enabled = true;
                Animation_timer_stalling_for_latency.AutoReset = false;
            }

        }
    }
    bool display_conditions_met()
    {
        return Showing == true && Observable_screenSize_reference != 0 && window_innerHeight != 0 && window_innerWidth != 0;
    }
}
@code{
    protected override Task OnParametersSetAsync()
    {
        if(Observable_screenSize_reference!=0)
        {
            Initialize_dimension_after_first_render();
            //start animation timer
        }
        return base.OnParametersSetAsync();
    }
}
@code{
    bool animation_signal = false;
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)//after first render
        {
            //don't use StateHasChanged in these processors/functions
            //Destruction_after_duration();
            //StateHasChanged();
        }
        else
        {
            if (display_conditions_met())
            {
                if(animation_signal==false)
                {
                    animation_signal = true;//skip this frame
                    StateHasChanged();//let's do the next frame (no more skipping)
                }
                else
                {
                    Destruction_after_duration();
                    Animation_expansion();//trigger a timer 
                }
            }
        }
    }
}
@if (display_conditions_met())
{
    <div style="height:100vh; width:100vw; position:fixed; background-color:#171717;">
        <div id="@element_Id.get(0)" style="@Layout_algorithm.Centering_element.By_bounds((Observable_screenSize_reference * 0.2F),(Observable_screenSize_reference * 0.2F),window_innerWidth,window_innerHeight, enabled_) height:@Initial_height; width:@Initial_width; transition:all @(((float)Duration/1000).ToString())s @(Material.css_styles.Curves_collection(1));background-color:red;">
        </div>
    </div>
}
```
